crypto/include/internal/asn1_dsa.h

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef HEADER_ASN1_DSA_H
+# define HEADER_ASN1_DSA_H
+
+/*
+ * Outputs the DER encoding of a DSA-Sig-Value or ECDSA-Sig-Value to *ppout
+ * and increments *ppout past the data just written.
+ * If either ppout or *ppout is NULL then the output size is calculated but
+ * no output is produced.
+ * Returns the number of bytes of output (or that would be output) or 0 on
+ * failure.
+ */
+size_t i2d_dsa_sig(const BIGNUM *r, const BIGNUM *s, unsigned char **ppout);
+
+/*
+ * Decodes a DSA-Sig-Value or ECDSA-Sig-Value from *ppin, which must be DER
+ * encoded, and increments *ppin past the data that was consumed.
+ * An attempt to consume more than len bytes results in failure.
+ * Returns 1 on success, 0 on failure.
+ */
+int d2i_dsa_sig(BIGNUM *r, BIGNUM *s, const unsigned char **ppin, size_t len);
```

Comment:
It returns -1?

Reply:
Actually, it is i2d_TYPE() that returns -1 on error.
It did return 1 on success (not -1), but I have made some changes.
The function is now called decode_der_dsa_sig() and it now returns a size_t: the number of bytes processed on input, or zero in the case of an error. No more i2d_*() or d2i_*(), since these new functions didn't really conform to the i2d/d2i way of doing things anyway.